test/performance/broker-latency/main.go

```diff
@@ -0,0 +1,286 @@
+
+	// The interval and timeout used for polling in checking event states.
+	pollInterval = 1 * time.Second
+	pollTimeout  = 4 * time.Minute
+)
+
+// flags for the image
+var (
+	benchmark          string
+	sinkURL            string
+	msgSize            int
+	eventNum           int
+	errorRateThreshold float64
+	timeout            int
+	encoding           string
+	eventTimeMap       map[string]chan time.Time
+	resultCh           chan state
+)
+
+// eventStatus is status of the event, for now only if all events are in received status can the
+// test be considered as PASS.
+type eventStatus int
+
+const (
+	sent eventStatus = iota
+	received
+	undelivered
+	dropped
+	duplicated
+	corrupted // TODO(Fredy-Z): corrupted status is not being used now
+)
+
+// state saves the data that is used to generate the metrics
+type state struct {
+	latency time.Duration
+	status  eventStatus
+}
+
+const (
+	makoAddr = "localhost:9813"
+)
+
+func init() {
+	flag.StringVar(&benchmark, "benchmark", "", "The mako benchmark ID")
+	flag.StringVar(&sinkURL, "sink", "http://default-broker", "The sink URL for the event destination.")
+	flag.IntVar(&msgSize, "msg-size", 100, "The size of each message we want to send. Generate random strings to avoid caching.")
+	flag.IntVar(&eventNum, "event-count", 10, "The number of events we want to send.")
+	flag.IntVar(&timeout, "timeout", 30, "Timeout in seconds. If we do not receive a message back within a limited time, we consider it to be dropped.")
+	flag.Float64Var(&errorRateThreshold, "error-rate-threshold", 0.1, "Rate of error event deliveries we allow. We fail the test if the error rate crosses the threshold.")
+	flag.StringVar(&encoding, "encoding", "binary", "The encoding of the cloud event, one of(binary, structured).")
+}
+
+func ts(t time.Time) float64 {
+	return float64(t.UnixNano()) / (1000.0 * 1000.0)
+}
+
+func main() {
+	// parse the command line flags
+	flag.Parse()
+
+	if benchmark == "" {
+		log.Fatal("-benchmark is a required flag.")
+	}
+
+	// get encoding
+	var encodingOption http.Option
+	switch encoding {
+	case "binary":
+		encodingOption = cloudevents.WithBinaryEncoding()
+	case "structured":
+		encodingOption = cloudevents.WithStructuredEncoding()
+	default:
+		failTest(fmt.Sprintf("unsupported encoding option: %q\n", encoding))
+	}
+
+	// create cloudevents client
+	t, err := cloudevents.NewHTTPTransport(
+		cloudevents.WithTarget(sinkURL),
+		encodingOption,
+	)
+	if err != nil {
+		failTest(fmt.Sprintf("failed to create transport: %v\n", err))
+	}
+	c, err := cloudevents.NewClient(t,
+		cloudevents.WithTimeNow(),
+		cloudevents.WithUUIDs(),
+	)
+	if err != nil {
+		failTest(fmt.Sprintf("failed to create client: %v\n", err))
+	}
+
+	ctx := context.Background()
+	q, qclose, err := quickstore.NewAtAddress(ctx, &qpb.QuickstoreInput{
+		BenchmarkKey: proto.String(benchmark),
+	}, makoAddr)
+	if err != nil {
+		failTest(fmt.Sprintf("failed NewAtAddress: %v\n", err))
+	}
+	defer qclose(ctx)
```

Comment:
You should be really careful here because if someone adds `context.WithTimeout()` above, you'll stop killing the sidecar and leak Job pods on-cluster.